src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -0,0 +1,617 @@
+    frontLeft.periodic();
+    frontRight.periodic();
+    rearLeft.periodic();
+    rearRight.periodic();
+    odometry.update(Rotation2d.fromDegrees(gyro.getYaw().getValue()), getModulePositions());
+  }
+
+  public SwerveModulePosition[] getModulePositions() {
+    return new SwerveModulePosition[] {
+        frontLeft.getPosition(),
+        frontRight.getPosition(),
+        rearLeft.getPosition(),
+        rearRight.getPosition()
+    };
+  };
+
+  public void burnFlashSparks() {
+    frontLeft.burnFlashSparks();
+    frontRight.burnFlashSparks();
+    rearLeft.burnFlashSparks();
+    rearRight.burnFlashSparks();
+  }
+
+  /**
+   * Returns the currently-estimated pose of the robot.
+   *
+   * @return The pose.
+   */
+  public Pose2d getPose() {
+    return odometry.getPoseMeters();
+  }
+
+  public ChassisSpeeds getChassisSpeeds() {
+    return lastSetChassisSpeeds;
+  }
+
+  /**
+   * applies the robot-relative output speeds of the FollowPathHolonomic Command
+   * adapted from drive() (but doesn't need the code to turn xSpeed, ySpeed, rot,
+   * fieldRelative into a ChassisSpeeds object)
+   * does not call the correctForDynamics method because we pass in a corrected
+   * lastSetChassisSpeeds (// TODO is this right???)
+   */
+  public void setOutputRobotRelativeSpeeds(ChassisSpeeds desiredChassisSpeeds) {
+    lastSetChassisSpeeds = desiredChassisSpeeds;
+    var swerveModuleStates = Constants.SwerveDrive.DRIVE_KINEMATICS.toSwerveModuleStates(desiredChassisSpeeds);
+    setModuleStates(swerveModuleStates);
+  }
+
+  /**
+   * Resets the odometry to the specified pose.
+   *
+   * @param pose The pose to which to set the odometry.
+   */
+  public void resetOdometry(Pose2d pose) {
+    // Just update the translation, not the yaw
+    Pose2d resetPose = new Pose2d(pose.getTranslation(), Rotation2d.fromDegrees(gyro.getYaw().getValue()));
+    odometry.resetPosition(Rotation2d.fromDegrees(gyro.getYaw().getValue()), getModulePositions(), resetPose);
+  }
+
+  public void resetYawToAngle(double yawDeg) {
+    double curYawDeg = gyro.getYaw().getValue();
+    double offsetToTargetDeg = targetHeadingDegrees - curYawDeg;
+    gyro.setYaw(yawDeg);
+    Pose2d curPose = getPose();
+    Pose2d resetPose = new Pose2d(curPose.getTranslation(), Rotation2d.fromDegrees(yawDeg));
+    odometry.resetPosition(Rotation2d.fromDegrees(yawDeg), getModulePositions(), resetPose);
+    targetHeadingDegrees = yawDeg + offsetToTargetDeg;
+  }
+
+  public void resetYaw() {
+    resetYawToAngle(0.0);
+  }
+
+  /**
+   * Correction for swerve second order dynamics issue. Borrowed from 254:
+   * https://github.com/Team254/FRC-2022-Public/blob/main/src/main/java/com/team254/frc2022/subsystems/Drive.java#L325
+   * Discussion:
+   * https://www.chiefdelphi.com/t/whitepaper-swerve-drive-skew-and-second-order-kinematics/416964
+   */
+  private static ChassisSpeeds correctForDynamics(ChassisSpeeds originalSpeeds) {
+    final double LOOP_TIME_S = 0.02;
```

Comment:
There was some discussion about arbitrarily making this larger helping, can we throw in a TODO to test that?